src/index.ts

```diff
@@ -1,37 +1,119 @@
+const isUnionType = ({ type }: t.ContextEntry) => type instanceof t.UnionType;
 
-export const formatValidationError = (error: t.ValidationError) => {
-    const path = error.context
+const jsToString = (value: t.mixed) =>
+    value === undefined ? 'undefined' : JSON.stringify(value);
+
+const keyPath = (ctx: t.Context) =>
+    ctx
         .map(c => c.key)
-        // The context entry with an empty key is the original type ("default
-        // context"), not an type error.
+        // The context entry with an empty key is the original
+        // type ("default context"), not an type error.
         .filter(key => key.length > 0)
         .join('.');
 
-    // The actual error is last in context
-    const maybeErrorContext = array.last(
-        // https://github.com/gcanti/fp-ts/pull/544/files
-        error.context as Array<t.ContextEntry>
+// The actual error is last in context
+const getErrorFromCtx = (validation: t.ValidationError) =>
+    // https://github.com/gcanti/fp-ts/pull/544/files
+    A.last(validation.context as Array<t.ContextEntry>);
+
+const getValidationContext = (validation: t.ValidationError) =>
+    // https://github.com/gcanti/fp-ts/pull/544/files
+    validation.context as Array<t.ContextEntry>;
+
+const expectedTypesToString = flow(
+    A.map(({ type }: t.ContextEntry) => `    ${type.name}`),
+    arr => arr.join('\n'),
+);
+
+export const formatUnionError = (
+    path: string,
+    errors: NEA.NonEmptyArray<t.ValidationError>,
+) => {
+    const expectedTypes = pipe(
+        errors,
+        A.map(getValidationContext),
+        A.map(ctx =>
+            pipe(
+                ctx,
+                // find the union type in the list of ContextEntry
+                A.findIndex(isUnionType),
+                // the next ContextEntry should be the
+                // type of this branch of the union
+                O.chain(n => A.lookup(n + 1, ctx)),
+            ),
+        ),
+        A.compact,
+    );
+
+    const value = pipe(
+        expectedTypes,
+        A.head,
+        O.map(v => v.actual),
+        O.getOrElse<unknown>(() => undefined),
     );
 
-    return pipe(maybeErrorContext, map(errorContext => {
-        const expectedType = errorContext.type.name;
-        return (
-            // https://github.com/elm-lang/core/blob/18c9e84e975ed22649888bfad15d1efdb0128ab2/src/Native/Json.js#L199
-            // tslint:disable-next-line:prefer-template
-            `Expecting ${expectedType}`
-                + (path === '' ? '' : ` at ${path}`)
-                + ` but instead got: ${jsToString(error.value)}.`
-        );
-    }))
+    const expected = expectedTypesToString(expectedTypes);
+
+    return expected.trim() === ''
+        ? O.none
+        : O.some(
+              // https://github.com/elm-lang/core/blob/18c9e84e975ed22649888bfad15d1efdb0128ab2/src/Native/Json.js#L199
+              // tslint:disable-next-line:prefer-template
+              `Expecting one of:\n${expected}` +
```

Comment:
one thing I'd change with your PR is that now there's two different places where the same error message is being constructed, either find a way for both code paths to merge, or extract the error message code to its own helper

Reply:
that and having two places to construct the same message seems like a likely way for them to go out of sync and introduce regressions, etc.

if my error message PR doesn't go through and this reporter chooses not to show error messages, then this particular problem isn't important, though not suppressing error messages seems like a good thing to me

Reply:
> but ideally if there's one "dominant" error like mentioned in #5 to make it choose a branch smartly, that one would probably be the error message I'd display.

Totally agree on that.

And yes, we should unify the way error messages are generated.